hazelcast/src/main/java/com/hazelcast/map/impl/event/QueryCacheNaturalFilteringStrategy.java

```diff
@@ -0,0 +1,250 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl.event;
+
+import com.hazelcast.core.EntryEventType;
+import com.hazelcast.map.impl.EntryEventFilter;
+import com.hazelcast.map.impl.EventListenerFilter;
+import com.hazelcast.map.impl.MapPartitionLostEventFilter;
+import com.hazelcast.map.impl.MapServiceContext;
+import com.hazelcast.map.impl.query.QueryEventFilter;
+import com.hazelcast.nio.Address;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.spi.EventFilter;
+import com.hazelcast.spi.impl.eventservice.impl.TrueEventFilter;
+import com.hazelcast.spi.serialization.SerializationService;
+import com.hazelcast.util.collection.Int2ObjectHashMap;
+
+import java.util.Collection;
+import java.util.Map;
+
+import static com.hazelcast.core.EntryEventType.ADDED;
+import static com.hazelcast.core.EntryEventType.EVICTED;
+import static com.hazelcast.core.EntryEventType.EXPIRED;
+import static com.hazelcast.core.EntryEventType.REMOVED;
+import static com.hazelcast.core.EntryEventType.UPDATED;
+
+/**
+ * A filtering strategy that preserves the default behavior in most cases, but processes entry events for listeners with
+ * predicates to fit with "query cache" concept. For example when the original event indicates an update from an old value that
+ * does not match the predicate to a new value that does match, then the entry listener will be notified with an ADDED event.
+ * This affects only map listeners with predicates and the way entry updates are handled. Put/remove operations are not
+ * affected, neither are listeners without predicates.
+ * This filtering strategy is used when Hazelcast property {@code hazelcast.map.entry.filtering.natural.event.types} is set
+ * to {@code true}. The complete decision matrix for event types published with this filtering strategy.
+ * <p>
+ * <table>
+ *     <tr>
+ *         <td>Old value</td>
+ *         <td>New value</td>
+ *         <td>Published event type</td>
+ *     </tr>
+ *     <tr>
+ *         <td>Match</td>
+ *         <td>Mismatch</td>
+ *         <td>REMOVED</td>
+ *     </tr>
+ *     <tr>
+ *         <td>Match</td>
+ *         <td>Match</td>
+ *         <td>UPDATED</td>
+ *     </tr>
+ *     <tr>
+ *         <td>Mismatch</td>
+ *         <td>Mismatch</td>
+ *         <td>NO MATCH</td>
+ *     </tr>
+ *     <tr>
+ *         <td>Mismatch</td>
+ *         <td>Match</td>
+ *         <td>ADDED</td>
+ *     </tr>
+ * </table>
```

Comment:
Would be great to have such a behaviour-describing table in the `DefaultEntryEventFilteringStrategy` so that one can compare.